nibabel/cifti2/cifti2_axes.py

```diff
@@ -0,0 +1,1224 @@
+import numpy as np
+from nibabel.cifti2 import cifti2
+from six import string_types
+from operator import xor
+
+
+def from_mapping(mim):
+    """
+    Parses the MatrixIndicesMap to find the appropriate CIFTI axis describing the rows or columns
+
+    Parameters
+    ----------
+    mim : cifti2.Cifti2MatrixIndicesMap
+
+    Returns
+    -------
+    subtype of Axis
+    """
+    return_type = {'CIFTI_INDEX_TYPE_SCALARS': Scalar,
+                   'CIFTI_INDEX_TYPE_LABELS': Label,
+                   'CIFTI_INDEX_TYPE_SERIES': Series,
+                   'CIFTI_INDEX_TYPE_BRAIN_MODELS': BrainModel,
+                   'CIFTI_INDEX_TYPE_PARCELS': Parcels}
+    return return_type[mim.indices_map_to_data_type].from_mapping(mim)
+
+
+def to_header(axes):
+    """
+    Converts the axes describing the rows/columns of a CIFTI vector/matrix to a Cifti2Header
+
+    Parameters
+    ----------
+    axes : iterable[Axis]
+        one or more axes describing each dimension in turn
+
+    Returns
+    -------
+    cifti2.Cifti2Header
+    """
+    axes = list(axes)
+    mims_all = []
+    matrix = cifti2.Cifti2Matrix()
+    for dim, ax in enumerate(axes):
+        if ax in axes[:dim]:
+            dim_prev = axes.index(ax)
+            mims_all[dim_prev].applies_to_matrix_dimension.append(dim)
+            mims_all.append(mims_all[dim_prev])
+        else:
+            mim = ax.to_mapping(dim)
+            mims_all.append(mim)
+            matrix.append(mim)
+    return cifti2.Cifti2Header(matrix)
+
+
+class Axis(object):
+    """
+    Generic object describing the rows or columns of a CIFTI vector/matrix
+
+    Attributes
+    ----------
+    arr : np.ndarray
+        (N, ) typed array with the actual information on each row/column
```

Comment:
@MichielCottaar  This documentation is a bit unclear. This corresponds more to an abstract class than to a generic object, which, at least to me, makes me think of Generic as in type-parameterised objects.

Reply:
The typed array has been removed and the constructors updated